src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs

```diff
@@ -18,6 +18,35 @@ public partial class MemoryMappedFile
             HandleInheritability inheritability, MemoryMappedFileAccess access,
             MemoryMappedFileOptions options, long capacity)
         {
+            bool isRegularFile = true;
+            Interop.Sys.FileStatus status = default;
+
+            if (fileStream != null)
+            {
+                int result = Interop.Sys.FStat(fileStream.SafeFileHandle, out status);
+                if (result != 0)
+                {
+                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+                    throw Interop.GetExceptionForIoErrno(errorInfo);
+                }
+
+                isRegularFile = (status.Mode & Interop.Sys.FileTypes.S_IFCHR) == 0;
+
+                // perform deferred argument check.
+                if (isRegularFile)
+                {
+                    if (access == MemoryMappedFileAccess.Read && capacity > status.Size)
+                    {
+                        throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
+                    }
+
+                    if (access == MemoryMappedFileAccess.Write)
+                    {
+                        throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
+                    }
+                }
+            }
+
             if (mapName != null)
```

Comment:
Should we pull this verification to the top of this method? There's no point in doing anything if the user tried to pass a mapName.

Reply:
If mapName is not null we can throw much earlier, instead of doing all the checks inside `VerifyMemoryMappedFileAccess`. What do you think, @wfurt?

Reply:
It seems like before we check MemoryMappedFileAccess.Write, then capacity > fileStream.Length, then mapName != null and then we may try fileStream.SetLength. It seems like best option would be to preserver that order and perhaps do more VerifyMemoryMappedFileAccess(). 
It seems like the consensus is that it would be OK to reorder exceptions if multiple parameters are invalid. 
If so, we can do what ever we want. But perhaps move that check to VerifyMemoryMappedFileAccess() as well to keep as much of the validation together?

Reply:
Let's preserve the order. I don't think it affects much to do the mapName check before or after.